repl/repl.go

```diff
@@ -88,98 +87,70 @@ func rep(rl *readline.Instance, thread *starlark.Thread, globals starlark.String
 
 	thread.SetLocal("context", ctx)
 
-	rl.SetPrompt(">>> ")
-	line, err := rl.Readline()
-	if err != nil {
-		return err // may be ErrInterrupt
-	}
+	eof := false
 
-	if l := strings.TrimSpace(line); l == "" || l[0] == '#' {
-		return nil // blank or comment
+	// readline returns EOF, ErrInterrupted, or a line including "\n".
+	rl.SetPrompt(">>> ")
+	readline := func() ([]byte, error) {
+		line, err := rl.Readline()
+		rl.SetPrompt("... ")
+		if err != nil {
+			if err == io.EOF {
+				eof = true
+			}
+			return nil, err
+		}
+		return []byte(line + "\n"), nil
 	}
 
-	// If the line contains a well-formed expression, evaluate it.
-	if _, err := syntax.ParseExpr("<stdin>", line, 0); err == nil {
-		if v, err := starlark.Eval(thread, "<stdin>", line, globals); err != nil {
-			PrintError(err)
-		} else if v != starlark.None {
-			fmt.Println(v)
+	// parse
+	f, err := syntax.ParseCompoundStmt("<stdin>", readline)
+	if err != nil {
+		if eof {
+			return io.EOF
 		}
+		PrintError(err)
 		return nil
 	}
 
-	// If the input so far is a single load or assignment statement,
-	// execute it without waiting for a blank line.
-	if f, err := syntax.Parse("<stdin>", line, 0); err == nil && len(f.Stmts) == 1 {
-		switch f.Stmts[0].(type) {
-		case *syntax.AssignStmt, *syntax.LoadStmt:
-			// Execute it as a file.
-			if err := execFileNoFreeze(thread, line, globals); err != nil {
-				PrintError(err)
+	// Transform a single expression statement to "$ = expr".
```

Comment:
Is there a way to evaluate the expression directly? I imagine there may be other ways for `$` to be introduced as a symbol here, maybe by the embedding program. It's probably not going to cause trouble now, but it seems like something that could be tripped over in the future.

Reply:
Done, I added EvalExpr, which is like Eval without the parsing. Like FileProgram, it has the caveat that it mutates (resolves) the AST. But you're right, this is cleaner.